plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ParallelStep.java

```diff
@@ -176,6 +176,7 @@ private void checkAllDone(boolean stepFailed) {
                 } else {
                     handler.context.onSuccess(success);
                 }
+                // TODO should this also bodies.clear()?
```

Comment:
The originally observed error, which I was unable to reproduce exactly in a functional test, involved the `CpsBodyExecution` being held from `ParallelStepExecution.bodies`:

```
an exception which occurred:
	in object of type org.jenkinsci.plugins.workflow.cps.FlowHead
	in field org.jenkinsci.plugins.workflow.cps.CpsThread.head
	in object org.jenkinsci.plugins.workflow.cps.CpsThread@282f4605
	in object of type org.jenkinsci.plugins.workflow.cps.CpsThread
	in field org.jenkinsci.plugins.workflow.cps.CpsBodyExecution.thread
	in object org.jenkinsci.plugins.workflow.cps.CpsBodyExecution@6c50e364
	in object of type org.jenkinsci.plugins.workflow.cps.CpsBodyExecution
	in element at index [1] of size [3]
	in field org.jenkinsci.plugins.workflow.cps.steps.ParallelStepExecution.bodies
	in object org.jenkinsci.plugins.workflow.cps.steps.ParallelStepExecution@4416bb37
	in object of type org.jenkinsci.plugins.workflow.cps.steps.ParallelStepExecution
	in field org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler.stepExecution
	in object org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler@253b016b
	in object of type org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler
	in field org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler$Callback.handler
	in object org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler$Callback@1193a1ed
	in object of type org.jenkinsci.plugins.workflow.cps.steps.ParallelStep$ResultHandler$Callback
	in element at index [0] of size [1]
	in field org.jenkinsci.plugins.workflow.cps.CpsBodyExecution.callbacks
	in object org.jenkinsci.plugins.workflow.cps.CpsBodyExecution@4a46a38c
	in object of type org.jenkinsci.plugins.workflow.cps.CpsBodyExecution
	in field org.jenkinsci.plugins.workflow.cps.CpsBodyExecution$SuccessAdapter.this$0
	in object org.jenkinsci.plugins.workflow.cps.CpsBodyExecution$SuccessAdapter@3bf876f0
	in object of type org.jenkinsci.plugins.workflow.cps.CpsBodyExecution$SuccessAdapter
	in field com.cloudbees.groovy.cps.impl.CallEnv.returnAddress
	in object com.cloudbees.groovy.cps.impl.FunctionCallEnv@463508bd
	in object of type com.cloudbees.groovy.cps.impl.FunctionCallEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.TryBlockEnv@4aa5555f
	in object of type com.cloudbees.groovy.cps.impl.TryBlockEnv
	in field com.cloudbees.groovy.cps.impl.CallEnv.caller
	in object com.cloudbees.groovy.cps.impl.ClosureCallEnv@3500c544
	in object of type com.cloudbees.groovy.cps.impl.ClosureCallEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.BlockScopeEnv@4d376ae2
	in object of type com.cloudbees.groovy.cps.impl.BlockScopeEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.BlockScopeEnv@23a9e319
	in object of type com.cloudbees.groovy.cps.impl.BlockScopeEnv
	in field com.cloudbees.groovy.cps.impl.CpsClosureDef.capture
	in object com.cloudbees.groovy.cps.impl.CpsClosureDef@43c6459e
	in object of type com.cloudbees.groovy.cps.impl.CpsClosureDef
	in field com.cloudbees.groovy.cps.impl.CpsClosure.def
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@705f8247
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in object of type java.util.LinkedHashMap
	in object of type WorkflowScript
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@64986d4
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@6a940c2f
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@1b0f4e56
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@174eaaca
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@5dfbe1d1
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@17446b35
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@76d8ea75
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@1422169e
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@165588a2
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@227c449e
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@13c699e6
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@1754fbe6
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@2d7e6136
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@6d6698b3
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in field groovy.lang.Closure.delegate
	in object org.jenkinsci.plugins.workflow.cps.CpsClosure2@4ac82550
	in object of type org.jenkinsci.plugins.workflow.cps.CpsClosure2
	in object of type java.util.HashMap
	in field org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.closures
	in object org.jenkinsci.plugins.workflow.cps.CpsThreadGroup@7a736c24
	in object of type org.jenkinsci.plugins.workflow.cps.CpsThreadGroup
Caused: java.lang.IllegalStateException: FlowHead loading problem at deserialize: Null FlowHead with id … in execution …
	at org.jenkinsci.plugins.workflow.cps.FlowHead.readResolve(FlowHead.java:207)
	at …
```

In this case the `ParallelStepExecution` was complete, so there was no reason for it to still be holding onto its `bodies`. (There was no _good_ reason for `program.dat` to be holding onto the `ParallelStepExecution` itself; this was due to a forgotten `def`.)